feat: Exchange GitHub OIDC credentials for AWS federated credentials

akash1810 opened the pull request:
## What does this change?
<!-- A PR should have enough detail to be understandable far in the future. e.g what is the problem/why is the change needed, how does it solve it and any questions or points of discussion. Prefer copying information from a Trello card over linking to it; the card may not always exist and reviewers may not have access to the board. -->

Following https://awsteele.com/blog/2021/09/15/aws-federation-comes-to-github-actions.html, produce an action that can be used to provide AWS federated credentials [via OIDC](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc.html). That is, allow GHA to interact with AWS w/out needing to supply GH with permanent AWS credentials 🎉 .

This allows one to move a project from TeamCity to GHA with relative ease. Currently, TeamCity has AWS credentials to upload to Riff-Raff's buckets via it's instance profile and projects are doing `sbt riffRaffUpload` to upload to Riff-Raff.

To migrate to GHA, one needs to:
  - Create a GHA workflow
  - Use this action to get AWS credentials
  - Run `sbt riffRaffUpload` as before

It's worth noting that this Action allows you to assume any AWS role, not just the one that allows you to upload to Riff-Raff's buckets. Pretty powerful stuff!

Whilst this feature is undocumented, it is on the official roadmap (https://github.com/github/roadmap/issues/249).

Some documentation has been added too. It's easiest to review [here](https://github.com/guardian/actions-assume-aws-role/blob/hackday/README.md).

akash1810 committed:
feat: Upload files to S3

akash1810 committed:
feat: Allow project name to be input

When not input, default to repository name.

sihil committed:
Simplify the action to only deal with creds

sihil committed:
Commit the built action artifacts

akash1810 committed:
fix: Produce a valid string to be written to the `GITHUB_ENV` file

akash1810 committed:
fix: Return raw value as string rather than `JSON.stringify` to remove the added quotes

akash1810 committed:
fix: Set old `AWS_REGION` env var too

It looks like this is needed by `WebIdentityTokenCredentialsProvider` in the Java SDKs.

akash1810 committed:
chore: Remove unused `wait` function

This came for free with the Action template repo

akash1810 commented on __tests__/main.test.ts:
```diff
@@ -1,24 +1,10 @@
 
 // shows how the runner will run a javascript action with env / stdout protocol
-test("test runs", () => {
+test.skip("test runs", () => {
```
This test runs the action locally and ensures it exits cleanly. I think a better version of this is an integration test, similar to https://github.com/guardian/actions-assume-aws-role/pull/6/files. Skipping for now.

akash1810 commented on package.json:
```diff
@@ -19,11 +19,13 @@
   "dependencies": {
-    "@actions/core": "^1.5.0"
+    "@actions/core": "^1.5.0",
+    "node-fetch": "2.6.2"
```
The latest version of `node-fetch` is `3.0.0` and comes included with type declarations. However it has also moved to ESM format and as we're compiling to CommonJS, we'd have to do a bit of [dancing](https://github.com/node-fetch/node-fetch/blob/main/docs/v3-UPGRADE-GUIDE.md) to work with it.

It's simpler to stick to v2 for now, at the cost of needing `@types/node-fetch`.

akash1810 committed:
test: Add integration style test

akash1810 commented on .github/workflows/test.yml:
```diff
@@ -16,9 +16,13 @@ jobs:
         with:
-          milliseconds: 1000
+          awsRoleToAssume: ${{ secrets.GU_RIFF_RAFF_ROLE_ARN }}
+      - run: aws s3 cp README.md s3://riffraff-artifact/actions-assume-aws-role/ci/action.yml
```
Uploading random files to Riff-Raff's bucket is totally fine, right?!

akash1810 replied:
Should the bucket name be stored in a repo secret?

jacobwinch replied:
> Should the bucket name be stored in a repo secret?

Yes, I think that would be preferable.

akash1810 commented:
Here's an example of the Action being used https://github.com/guardian/prism/pull/259.

jacobwinch commented on src/main.ts:
```diff
@@ -1,16 +1,70 @@
+    };
+
+    core.debug("*** START config ***");
+    core.debug(JSON.stringify(config));
```
IIUC debug logs will be [hidden by default](https://github.com/actions/toolkit/tree/main/packages/core#logging), but users can [choose to enable these log lines manually](https://github.com/actions/toolkit/blob/main/docs/action-debugging.md#how-to-access-step-debug-logs). Should we remove this debug logging before publishing to avoid allowing people to accidentally leak the `awsRoleToAssume` or `ACTIONS_ID_TOKEN_REQUEST_TOKEN`?

akash1810 replied:
I think any token or secret-like strings were automatically redacted. We added these to help debugging, but we've proven the process works now so I think we can delete these lines.

jacobwinch commented on src/main.ts:
```diff
@@ -1,16 +1,70 @@
-    await wait(parseInt(ms, 10));
-    core.debug(new Date().toTimeString());
+    core.debug("*** START environment ***");
+    core.debug(envString);
```
I also wonder if this should be removed

akash1810 replied:
Done.

akash1810 committed:
chore: Remove debugging statements

We no longer need these and there's a risk that they may leak information unintentionally.

jacobwinch reviewed (dismissed):
Great work 💯 🎉 🚢

akash1810 committed:
fix: Create token file within the tmp directory of the OS

akash1810 committed:
refactor: rename for clarity

akash1810 committed:
chore: Update README

akash1810 committed:
chore: v1.0.0

jacobwinch commented on README.md:
```diff
@@ -1,105 +1,146 @@
+[This blog][awsteele] provides a lot of detail.
 
-This template includes compilation support, tests, a validation workflow, publishing, and versioning guidance.  
+In short, you first need to create three resources:
```
Is it correct to say that Guardian projects which just want to upload artifacts to Riff-Raff S3 buckets will be able to skip this step? That is, are we all able to share a single policy/role/OIDC provider in the Deploy Tools account?

If so, we might want to make a note of this to avoid confusion.

jacobwinch commented on README.md:
```diff
@@ -1,105 +1,146 @@
-
-You can now validate the action by referencing `./` in a workflow in your repo (see [test.yml](.github/workflows/test.yml))
+## Usage
+1. Add the `AWS::IAM::Role` ARN as a GitHub Actions [secret] (we'll assume the secret has been called `GU_ACTIONS_ROLE`)
```
I think steps 1 & 2 might be taken care of for Guardian projects? If so, perhaps we can make another note about that?

akash1810 committed:
chore: Reduce the steps needed for Guardian repositories
